cmd/erasure-multipart.go

```diff
@@ -347,6 +347,12 @@ func (er erasureObjects) CopyObjectPart(ctx context.Context, srcBucket, srcObjec
 //
 // Implements S3 compatible Upload Part API.
 func (er erasureObjects) PutObjectPart(ctx context.Context, bucket, object, uploadID string, partID int, r *PutObjReader, opts ObjectOptions) (pi PartInfo, e error) {
+	uploadIDLock := er.NewNSLock(ctx, bucket, pathJoin(object, uploadID))
```

Comment:
Shouldn't this be a read lock? Otherwise it doesn't seem possible to upload parts concurrently.

If it is to protect against concurrently uploading the same part the part ID should be added to the lock name.

Maybe you need locks both for each part and the global state?

Reply:
This is how it is today @klauspost not changing that, concurrently uploading is not safe.

Reply:
You can't upload different parts concurrently? Sounds like a bad restriction.

Reply:
What is the problem grabbing a read lock here?

Reply:
In the past, we had many variations with lots of fine-grained approaches 

- We need to update a single xl.meta file for each part (we need to serialize all PutObjectPart()) calls here
- We can allow uploading content in parallel

Reply:
@harshavardhana Let's revisit this in the future. It should be possible to upload multiple parts in parallel. Designing parts to be independent and only collecting the information when finalizing should definitely be feasible.

Reply:
I was able to fix it @klauspost